### PR TITLE
add avx512 instrinsic

### PR DESCRIPTION
**Pull request by fingera**

### CODE REVIEW

**Comment by fingera** [comment]
ci error: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79127

**Comment by fingera** [comment]
Need to rebase?

**Comment by fingera** [comment]
because avx512 added _mm512_rol_epi32 instrisic, may faster in the future.
can i do something for merge

File: configure.ac [hunk 1]
```diff
@@ -375,6 +376,27 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
 )
 CXXFLAGS="$TEMP_CXXFLAGS"
 
+case $host in
+  *mingw*)
```
Comment (luke-jr) [inline configure.ac]: Why is this disallowed on mingw? Add a comment...

Comment (fingera) [reply configure.ac]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79127

Comment (luke-jr) [reply configure.ac]: Shouldn't your below test cleanly fail in that scenario?

Comment (fingera) [reply configure.ac]: Most servers are linux and support avx512?

Comment (luke-jr) [reply configure.ac]: That's irrelevant...

Comment (fingera) [reply configure.ac]: How do I do？ update ci mingw version?

Comment (luke-jr) [reply configure.ac]: You shouldn't need to.

File: configure.ac [hunk 2]
```diff
@@ -375,6 +376,27 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
...
+    ;;
+  *)
+    TEMP_CXXFLAGS="$CXXFLAGS"
+    CXXFLAGS="$CXXFLAGS $AVX512_CXXFLAGS"
+    AC_MSG_CHECKING(for AVX512 intrinsics)
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
```
Comment (luke-jr) [inline configure.ac]: Use `AC_LINK_IFELSE` here instead

Comment (fingera) [reply configure.ac]: Add a little code to make mingw generate xmm register more than 16 is ok? @luke-jr

File: src/crypto/sha256_avx512.cpp [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+__m512i inline Add(__m512i x, __m512i y, __m512i z, __m512i w) { return Add(Add(x, y), Add(z, w)); }
+__m512i inline Add(__m512i x, __m512i y, __m512i z, __m512i w, __m512i v) { return Add(Add(x, y, z), Add(w, v)); }
+__m512i inline Inc(__m512i& x, __m512i y) { x = Add(x, y); return x; }
+__m512i inline Inc(__m512i& x, __m512i y, __m512i z) { x = Add(x, y, z); return x; }
+__m512i inline Inc(__m512i& x, __m512i y, __m512i z, __m512i w) { x = Add(x, y, z, w); return x; }
+__m512i inline Xor(__m512i x, __m512i y) { return _mm512_xor_si512(x, y); }
+__m512i inline Xor(__m512i x, __m512i y, __m512i z) { return Xor(Xor(x, y), z); }
+__m512i inline Or(__m512i x, __m512i y) { return _mm512_or_si512(x, y); }
+__m512i inline And(__m512i x, __m512i y) { return _mm512_and_si512(x, y); }
+__m512i inline ShR(__m512i x, int n) { return _mm512_srli_epi32(x, n); }
```
Comment (practicalswift) [inline src/crypto/sha256_avx512.cpp]: The sign of the result may change here due to the conversion?

**Comment by dongcarl** [comment]
@fingera Would you like to rebase and keep working on this? I'd be happy to rebase for you.

**Comment by fingera** [comment]
> @fingera Would you like to rebase and keep working on this? I'd be happy to rebase for you.

Thankyou. Just wait for ci-mingw updates :)

**Comment by dongcarl** [comment]
@fingera If I understand you correctly, the current `AC_LANG_PROGRAM` [here](https://github.com/bitcoin/bitcoin/pull/13989/files/2fa6fa40c34fea86ebd947ccf20da5a3049e5e62#diff-67e997bcfdac55191033d57a16d1408aR386) will not fail if `$host` is mingw. But if we use your longer `AC_LANG_PROGRAM` from https://github.com/fingera/bitcoin/commit/585a8c8b5b500ca8cb2bab4e2d231895a6298703, it will fail just like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79127 says?

**Comment by fingera** [comment]
> @fingera If I understand you correctly, the current `AC_LANG_PROGRAM` [here](https://github.com/bitcoin/bitcoin/pull/13989/files/2fa6fa40c34fea86ebd947ccf20da5a3049e5e62#diff-67e997bcfdac55191033d57a16d1408aR386) will not fail if `$host` is mingw. But if we use your longer `AC_LANG_PROGRAM` from [fingera@585a8c8](https://github.com/fingera/bitcoin/commit/585a8c8b5b500ca8cb2bab4e2d231895a6298703), it will fail just like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79127 says?

Yes. But it’s too long, I think disallowed is better

**Comment by jamesob** [comment]
Per microbenchmarks, this change appears to have some pretty substantial performance improvements.

I've rebased this PR and run the microbenches on an Intel(R) Xeon(R) Silver 4116 CPU @ 2.10GHz. Notably, `micro.gcc.MerkleRoot` and `micro.gcc.SHA256D64_1024` are, respectively, 1.76x and 1.91x slower on master relative to this branch.

Kind of curious that we only see performance improvements on gcc though - is that expected?

Benchmarking data follows, significant results bolded.

### review/1-avx512 vs. master (relative)
|              bench name               |  x  | review/1-avx512 | master |
|---------------------------------------|----:|----------------:|-------:|
| build.make.22.clang.total_secs        |   1 |           1.000 |  1.003 |
| build.make.22.clang.peak_rss_KiB      |   1 |           1.000 |  1.000 |
| micro.clang.j=4.total_secs            |   3 |           1.007 |  1.000 |
| micro.clang.MerkleRoot.total_secs     |   3 |           1.008 |  1.000 |
| micro.clang.SHA1.total_secs           |   3 |           1.006 |  1.000 |
| micro.clang.SHA256.total_secs         |   3 |           1.000 |  1.003 |
| micro.clang.SHA256D64_1024.total_secs |   3 |           1.001 |  1.000 |
| micro.clang.SHA256_32b.total_secs     |   3 |           1.000 |  1.000 |
| micro.clang.SHA512.total_secs         |   3 |           1.000 |  1.000 |
| micro.gcc.j=4.total_secs              |   3 |           1.000 |  1.164 |
| **micro.gcc.MerkleRoot.total_secs**       |   3 |           1.000 |  1.760 |
| micro.gcc.SHA1.total_secs             |   3 |           1.001 |  1.000 |
| micro.gcc.SHA256.total_secs           |   3 |           1.008 |  1.000 |
| **micro.gcc.SHA256D64_1024.total_secs**   |   3 |           1.000 |  1.912 |
| micro.gcc.SHA256_32b.total_secs       |   3 |           1.002 |  1.000 |
| micro.gcc.SHA512.total_secs           |   3 |           1.000 |  1.003 |

### review/1-avx512 vs. master (absolute)
|              bench name               |  x  |    review/1-avx512     |         master         |
|---------------------------------------|----:|------------------------|------------------------|
| build.make.22.clang.total_secs        |   1 | 127.3094 (± 0.0000)    | 127.6333 (± 0.0000)    |
| build.make.22.clang.peak_rss_KiB      |   1 | 593740.0000 (± 0.0000) | 593512.0000 (± 0.0000) |
| micro.clang.j=4.total_secs            |   3 | 48.1295 (± 0.4233)     | 47.7760 (± 0.0253)     |
| micro.clang.MerkleRoot.total_secs     |   3 | 0.0019 (± 0.0000)      | 0.0019 (± 0.0000)      |
| micro.clang.SHA1.total_secs           |   3 | 0.0029 (± 0.0000)      | 0.0029 (± 0.0000)      |
| micro.clang.SHA256.total_secs         |   3 | 0.0049 (± 0.0000)      | 0.0049 (± 0.0000)      |
| micro.clang.SHA256D64_1024.total_secs |   3 | 0.0002 (± 0.0000)      | 0.0002 (± 0.0000)      |
| micro.clang.SHA256_32b.total_secs     |   3 | 0.0000 (± 0.0000)      | 0.0000 (± 0.0000)      |
| micro.clang.SHA512.total_secs         |   3 | 0.0045 (± 0.0000)      | 0.0045 (± 0.0000)      |
| micro.gcc.j=4.total_secs              |   3 | 42.0600 (± 0.1458)     | 48.9681 (± 0.1150)     |
| **micro.gcc.MerkleRoot.total_secs**       |   3 | 0.0011 (± 0.0000)      | 0.0020 (± 0.0000)      |
| micro.gcc.SHA1.total_secs             |   3 | 0.0029 (± 0.0000)      | 0.0029 (± 0.0000)      |
| micro.gcc.SHA256.total_secs           |   3 | 0.0049 (± 0.0001)      | 0.0049 (± 0.0000)      |
| **micro.gcc.SHA256D64_1024.total_secs**   |   3 | 0.0001 (± 0.0000)      | 0.0002 (± 0.0000)      |
| micro.gcc.SHA256_32b.total_secs       |   3 | 0.0000 (± 0.0000)      | 0.0000 (± 0.0000)      |
| micro.gcc.SHA512.total_secs           |   3 | 0.0048 (± 0.0000)      | 0.0048 (± 0.0000)      |

**Comment by dongcarl** [comment]
> > @fingera If I understand you correctly, the current `AC_LANG_PROGRAM` [here](https://github.com/bitcoin/bitcoin/pull/13989/files/2fa6fa40c34fea86ebd947ccf20da5a3049e5e62#diff-67e997bcfdac55191033d57a16d1408aR386) will not fail if `$host` is mingw. But if we use your longer `AC_LANG_PROGRAM` from [fingera@585a8c8](https://github.com/fingera/bitcoin/commit/585a8c8b5b500ca8cb2bab4e2d231895a6298703), it will fail just like [gcc.gnu.org/bugzilla/show_bug.cgi?id=79127](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79127) says?
> 
> Yes. But it’s too long, I think disallowed is better

The length of the `AC_LANG_PROGRAM` is completely fine, I believe we'd rather have accurate detection than broad-stroke disabling. Please rebase that branch over `master` and I'd be happy to take a look! :smile:

**Comment by jamesob** [comment]
Did some additional profiling and it's not clear to me that this change is worth pursuing.

Subsequent benching confirms that SHA-intensive microbenchmarks show improvement:

microbenches

However when sampling IBD performance (500_000 -> 510_000), there's only a minor difference in mean wall clock time taken:

ibd local range 500000 510000

**Comment by fingera** [comment]
> Did some additional profiling and it's not clear to me that this change is worth pursuing.
> 
> Subsequent benching confirms that SHA-intensive microbenchmarks show improvement:
> 
> microbenches
> 
> However when sampling IBD performance (500_000 -> 510_000), there's only a minor difference in mean wall clock time taken:
> 
> ibd local range 500000 510000

Because in this benchmark, merkle root 

> Did some additional profiling and it's not clear to me that this change is worth pursuing.
> 
> Subsequent benching confirms that SHA-intensive microbenchmarks show improvement:
> 
> microbenches
> 
> However when sampling IBD performance (500_000 -> 510_000), there's only a minor difference in mean wall clock time taken:
> 
> ibd local range 500000 510000

May be most of the time on diskIO？

File: src/crypto/sha256_avx512.cpp [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+__m512i inline Add(__m512i x, __m512i y, __m512i z, __m512i w) { return Add(Add(x, y), Add(z, w)); }
+__m512i inline Add(__m512i x, __m512i y, __m512i z, __m512i w, __m512i v) { return Add(Add(x, y, z), Add(w, v)); }
+__m512i inline Inc(__m512i& x, __m512i y) { x = Add(x, y); return x; }
+__m512i inline Inc(__m512i& x, __m512i y, __m512i z) { x = Add(x, y, z); return x; }
+__m512i inline Inc(__m512i& x, __m512i y, __m512i z, __m512i w) { x = Add(x, y, z, w); return x; }
+__m512i inline Xor(__m512i x, __m512i y) { return _mm512_xor_si512(x, y); }
+__m512i inline Xor(__m512i x, __m512i y, __m512i z) { return Xor(Xor(x, y), z); }
+__m512i inline Or(__m512i x, __m512i y) { return _mm512_or_si512(x, y); }
+__m512i inline And(__m512i x, __m512i y) { return _mm512_and_si512(x, y); }
+__m512i inline ShR(__m512i x, unsigned int n) { return _mm512_srli_epi32(x, n); }
```
Comment (practicalswift) [inline src/crypto/sha256_avx512.cpp]: There is an implicit conversion from `unsigned int` to `int` here. Could be made explicit to make it easier to reason about correctness?

Comment (fingera) [reply src/crypto/sha256_avx512.cpp]: gcc and Intel: __m512i _mm512_srli_epi32 (__m512i a, unsigned int imm8)
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm512_srli_epi32&expand=5515
https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/avx512fintrin.h
clang: __m512i _mm512_srli_epi32(__m512i __A, int __B)
https://clang.llvm.org/doxygen/avx512fintrin_8h_source.html#l05168

defined to macro?
#define ShR _mm512_srli_epi32

Comment (practicalswift) [reply src/crypto/sha256_avx512.cpp]: Oh, got it. Thanks for clarifying.

Also saw this old comment of mine: https://github.com/bitcoin/bitcoin/pull/13989#discussion_r221537873

Sorry for the confusion :-)

**Comment by fanquake** [comment]
@dongcarl Were you still interested in following up here?

**Review by dongcarl — CHANGES_REQUESTED** [review]
`AVX512` should be disabled if `SHA-NI` is enabled, like was done for `AVX2` here: https://github.com/bitcoin/bitcoin/blob/1b28bca04c2767c8bca21d66bd6978f358b0a96a/src/crypto/sha256.cpp#L613

See discussion here: http://www.erisian.com.au/bitcoin-core-dev/log-2019-06-26.html#l-511

**Comment by dongcarl** [comment]
- Code changes look good. Checked that it matches all instances of what we do for `AVX2`.
- Build system changes lightly tested on Linux machine w/o AVX512
  - Correctly enables AVX512 based on compiler compatibility, not system capability
  - Correctly disables AVX512 for gcc 4.8

-----

If someone has OSX/Windows machines, please test as well.

**Comment by fanquake** [comment]
> If someone has OSX/Windows machines, please test as well.

@dongcarl Is there anything in particular you'd like tested? I can test on macOS.

**Comment by dongcarl** [comment]
@fanquake Mostly if the build system changes work correctly. (I think AVX was added to clang in 3.9?)

**Comment by fanquake** [comment]
@dongcarl No worries.

Tested on `macOS` with `Clang`. Looks like support for AVX512 arrived in LLVM in [3.9](https://releases.llvm.org/3.9.0/docs/ReleaseNotes.html):
```bash
clang --version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
```

```bash
./configure | grep -i -E 'AVX'
checking whether C++ compiler accepts -mavx -mavx2... yes
checking whether C++ compiler accepts -mavx -mavx2 -mavx512f... yes
checking for AVX2 intrinsics... yes
checking for AVX512 intrinsics... yes
```

However my CPU does not support AVX512:
```bash
sysctl -a | grep machdep.cpu.features
AVX1.0
sysctl -a | grep machdep.cpu.leaf7_features
AVX2
```

Tested inside a `debian:jessie-slim` Docker container that AVX512 is disabled for GCC 4.8 and enabled for GCC 4.9 [(Intel AVX-512 support was added to GCC in 4.9)](https://gcc.gnu.org/gcc-4.9/changes.html).

GCC 4.8 - `gcc-4.8 (Debian 4.8.4-1) 4.8.4`:
```bash
./configure CC=gcc-4.8 CXX=g++-4.8 | grep -i 'AVX'
configure: WARNING: Doxygen not found
checking whether C++ compiler accepts -mavx -mavx2... yes
checking whether C++ compiler accepts -mavx -mavx2 -mavx512f... no
checking for AVX2 intrinsics... yes
checking for AVX512 intrinsics... no
```

GCC 4.9 - `gcc (Debian 4.9.2-10+deb8u2) 4.9.2`:
```bash
./configure --disable-wallet | grep -i 'AVX'
configure: WARNING: Doxygen not found
checking whether C++ compiler accepts -mavx -mavx2... yes
checking whether C++ compiler accepts -mavx -mavx2 -mavx512f... yes
checking for AVX2 intrinsics... yes
checking for AVX512 intrinsics... yes
```

Also tested with Clang inside the same box:

Clang 3.5 `Debian clang version 3.5.0-10 (tags/RELEASE_350/final) (based on LLVM 3.5.0)`:
```bash
./configure CC=clang-3.5 CXX=clang-3.5 | grep -i 'avx'
configure: WARNING: Doxygen not found
checking whether C++ compiler accepts -mavx -mavx2... yes
checking whether C++ compiler accepts -mavx -mavx2 -mavx512f... yes
checking for AVX2 intrinsics... yes
checking for AVX512 intrinsics... no
```

Clang 4.0 `clang version 4.0.1-10~deb8u1 (tags/RELEASE_401/final)`:
```bash
./configure --disable-wallet CC=clang-4.0 CXX=clang-4.0 | grep -i 'avx'
configure: WARNING: Doxygen not found
checking whether C++ compiler accepts -mavx -mavx2... yes
checking whether C++ compiler accepts -mavx -mavx2 -mavx512f... yes
checking for AVX2 intrinsics... yes
checking for AVX512 intrinsics... yes
```

It'd be great to have @sipa , @gmaxwell  or @theuni give this another look over.

**Comment by laanwj** [comment]
Concept ACK, I have no hardware to test this on.

**Comment by promag** [comment]
I also don't have hardware to test, concept ACK.

**Comment by fingera** [comment]
iMAC PRO or MAC PRO :)
Most cloud servers are already supported: https://aws.amazon.com/ec2/instance-types/
The popularity of desktops maybe wait for the future :<

File: src/crypto/sha256_avx512.cpp [hunk 5]
```diff
@@ -0,0 +1,350 @@
+#ifdef ENABLE_AVX512
+
+#include <stdint.h>
+#include <immintrin.h>
+
+#include <crypto/sha256.h>
```
Comment (theuni) [inline src/crypto/sha256_avx512.cpp]: Why?

**Review by theuni — COMMENTED** [review]
Concept ACK.

Build system changes look good at a glance. I can't test this either :(

**Comment by sipa** [comment]
Given the lack of common platforms where AVX512 occurs, and because IIRC on some of them it causes a clock speed reduction even for non-AVX512 instructions, it may not be wise to enable this by default at runtime. Does anyone have benchmarks that show otherwise?

A possibility is turning the autodetection code into an actual benchmark that's run at startup, finding which code is fastest for what empirically rather than guessing that X is always faster than Y if X is available.

File: configure.ac [hunk 6]
```diff
@@ -412,6 +413,87 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
 )
 CXXFLAGS="$TEMP_CXXFLAGS"
 
+TEMP_CXXFLAGS="$CXXFLAGS"
+CXXFLAGS="$CXXFLAGS $AVX512_CXXFLAGS"
+AC_MSG_CHECKING(for AVX512 intrinsics)
+AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+    #include <stdint.h>
```
[the remark this answers is not in the document]
Comment (fingera) [reply configure.ac]: A big program to trigger MINGW-GCC BUG

File: src/crypto/sha256_avx512.cpp (hunk 5 above)
Comment (fingera) [reply src/crypto/sha256_avx512.cpp]: Does not seem to be used sha256.h
Keep the same as other files(avx2 sse4)?

**Comment by fingera** [comment]
> Given the lack of common platforms where AVX512 occurs, and because IIRC on some of them it causes a clock speed reduction even for non-AVX512 instructions, it may not be wise to enable this by default at runtime. Does anyone have benchmarks that show otherwise?
> 
> A possibility is turning the autodetection code into an actual benchmark that's run at startup, finding which code is fastest for what empirically rather than guessing that X is always faster than Y if X is available.

I think all platforms AVX512 will be faster
avx2 has frequency scaling too

**Comment by sipa** [comment]
@fingera Maybe, seeing benchmarks would certainly be more convincing. Given that executing AVX512 instructions (as far as I know) may slow down other instructions, it may even be the case that in a microbenchmark AVX512-based SHA256 code is a win, but in realistic load conditions where SHA256 operations are interleaved with other things, it is not.

**Comment by fingera** [comment]
@sipa Yes your are right
https://en.wikichip.org/wiki/intel/frequency_behavior

**Comment by fingera** [comment]
https://github.com/travisdowns/avx-turbo
```
CPUID highest leaf  : [ dh]
Running as root     : [YES]
MSR reads supported : [NO ]
CPU pinning enabled : [YES]
CPU supports AVX2   : [YES]
CPU supports AVX-512: [YES]
CPUID doesn't support leaf 0x15, falling back to manual TSC calibration.
tsc_freq = 2499.9 MHz (from calibration loop)
CPU brand string: Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz
1 available CPUs: [0]
Can't use cpuid leaf 0xb to filter out hyperthreads, CPU too old or AMD
1 physical cores: [0]
Will test up to 1 CPUs
Cores | ID                  | Description                     | OVRLP1 | OVRLP2 | OVRLP3 | Mops
1     | pause_only          | pause instruction               |  1.000 |  1.000 | 1.000  | 19260
1     | ucomis              | SSE scalar ucomis loop          |  1.000 |  1.000 | 1.000  |  860
1     | ucomis_vex          | VEX scalar ucomis loop          |  1.000 |  1.000 | 1.000  |  546
1     | scalar_iadd         | Scalar integer adds             |  1.000 |  1.000 | 1.000  | 2700
1     | avx128_iadd         | 128-bit integer serial adds     |  1.000 |  1.000 | 1.000  | 2699
1     | avx256_iadd         | 256-bit integer serial adds     |  1.000 |  1.000 | 1.000  | 2700
1     | avx512_iadd         | 512-bit integer adds            |  1.000 |  1.000 | 1.000  | 2699
1     | avx128_iadd_t       | 128-bit integer parallel adds   |  1.000 |  1.000 | 1.000  | 8100
1     | avx256_iadd_t       | 256-bit integer parallel adds   |  1.000 |  1.000 | 1.000  | 8099
1     | avx128_mov_sparse   | 128-bit reg-reg mov             |  1.000 |  1.000 | 1.000  | 2700
1     | avx256_mov_sparse   | 256-bit reg-reg mov             |  1.000 |  1.000 | 1.000  | 2700
1     | avx512_mov_sparse   | 512-bit reg-reg mov             |  1.000 |  1.000 | 1.000  | 2700
1     | avx128_merge_sparse | 128-bit reg-reg merge mov       |  1.000 |  1.000 | 1.000  | 2700
1     | avx256_merge_sparse | 256-bit reg-reg merge mov       |  1.000 |  1.000 | 1.000  | 2700
1     | avx512_merge_sparse | 512-bit reg-reg merge mov       |  1.000 |  1.000 | 1.000  | 2700
1     | avx128_vshift       | 128-bit variable shift (vpsrld) |  1.000 |  1.000 | 1.000  | 2699
1     | avx256_vshift       | 256-bit variable shift (vpsrld) |  1.000 |  1.000 | 1.000  | 2699
1     | avx512_vshift       | 512-bit variable shift (vpsrld) |  1.000 |  1.000 | 1.000  | 2699
1     | avx128_vshift_t     | 128-bit variable shift (vpsrld) |  1.000 |  1.000 | 1.000  | 5398
1     | avx256_vshift_t     | 256-bit variable shift (vpsrld) |  1.000 |  1.000 | 1.000  | 5398
1     | avx512_vshift_t     | 512-bit variable shift (vpsrld) |  1.000 |  1.000 | 1.000  | 2699
1     | avx128_imul         | 128-bit integer muls            |  1.000 |  1.000 | 1.000  |  540
1     | avx256_imul         | 256-bit integer muls            |  1.000 |  1.000 | 1.000  |  540
1     | avx512_imul         | 512-bit integer muls            |  1.000 |  1.000 | 1.000  |  493
1     | avx128_fma_sparse   | 128-bit 64-bit sparse FMAs      |  1.000 |  1.000 | 1.000  | 2700
1     | avx256_fma_sparse   | 256-bit 64-bit sparse FMAs      |  1.000 |  1.000 | 1.000  | 2700
1     | avx512_fma_sparse   | 512-bit 64-bit sparse FMAs      |  1.000 |  1.000 | 1.000  | 2700
1     | avx128_fma          | 128-bit serial DP FMAs          |  1.000 |  1.000 | 1.000  |  675
1     | avx256_fma          | 256-bit serial DP FMAs          |  1.000 |  1.000 | 1.000  |  675
1     | avx512_fma          | 512-bit serial DP FMAs          |  1.000 |  1.000 | 1.000  |  675
1     | avx128_fma_t        | 128-bit parallel DP FMAs        |  1.000 |  1.000 | 1.000  | 5398
1     | avx256_fma_t        | 256-bit parallel DP FMAs        |  1.000 |  1.000 | 1.000  | 5398
1     | avx512_fma_t        | 512-bit parallel DP FMAs        |  1.000 |  1.000 | 1.000  | 4598
1     | avx512_vpermw       | 512-bit serial WORD permute     |  1.000 |  1.000 | 1.000  |  450
1     | avx512_vpermw_t     | 512-bit parallel WORD permute   |  1.000 |  1.000 | 1.000  | 1350
1     | avx512_vpermd       | 512-bit serial DWORD permute    |  1.000 |  1.000 | 1.000  |  900
1     | avx512_vpermd_t     | 512-bit parallel DWORD permute  |  1.000 |  1.000 | 1.000  | 2699
```

**Comment by fingera** [comment]
avx512_vshift_t speed: 50%
avx512_imul speed: 90%
Execution of complex instructions will cause the CPU clock upper and lower limits to drop at that time.
I think we didn't use these instructions.

**Comment by dongcarl** [comment]
@fingera Could you show benchmarks of the instructions we're using? I'm most interested in the `epi32` instructions.

**Comment by fanquake** [comment]
We're still waiting on benchmarks, PR comments need addressing, and the user that opened this seems to have disappeared from GitHub. Going to mark as up for grabs and close for now.

**Pull request closed without merging**